src/lang/proc_macros/controller.rs

```diff
@@ -152,16 +159,28 @@ impl ProcMacroClientController {
                     return;
                 };
 
-                let new_plugin_suite = proc_macro_plugin_suite(defined_macros);
-
-                // Store current plugins for identity comparison, so we can remove them if we
-                // restart proc-macro-server.
-                let previous_plugin_suite = self.plugin_suite.replace(new_plugin_suite.clone());
-
-                if let Some(previous_plugin_suite) = previous_plugin_suite {
-                    db.replace_plugin_suite(previous_plugin_suite, new_plugin_suite);
-                } else {
-                    db.add_plugin_suite(new_plugin_suite);
+                self.remove_current_plugins_from_db(db);
+
+                self.crate_plugin_suites = proc_macro_plugin_suites(defined_macros)
+                    .into_iter()
+                    .map(|(package_id, suite)| {
+                        // Here we rely on the implicit contract that keys of the
+                        // `DefinedMacrosResponse.macros_by_package_id` are of form
+                        // `PackageId.to_serialized_string()` which is always equal to
+                        // `scarb_metadata::CompilationUnitComponentId.repr` and has the form
+                        // "<crate-name> <version> (<source-path>)".
+                        let crate_name = package_id.split(' ').next().unwrap().to_smolstr();
+                        let crate_long_id =
+                            CrateLongId::Real { name: crate_name, discriminator: Some(package_id) };
+
+                        (crate_long_id, suite)
+                    })
+                    .collect::<OrderedHashMap<_, _>>();
+
+                for (crate_long_id, plugin_suite) in self.crate_plugin_suites.clone() {
```

Comment:
[nit] Clone only relevant items, not whole map.
```suggestion
                for (crate_long_id, plugin_suite) in &self.crate_plugin_suites {
```